libraries/platforms/dromajo-vcs/bsg_manycore_platform.cpp

```diff
@@ -0,0 +1,495 @@
+    return HB_MC_INVALID;
+  }
+
+  // Host doesn't send responses to the mamycore
+  if (type == HB_MC_FIFO_TX_RSP)
+  {
+    manycore_pr_err(mc, "TX Response Not Supported!\n");
+    return HB_MC_NOIMPL;
+  }
+
+  int credits_used = 0;
+  do {
+    err = bp_hb_get_credits_used(&credits_used);
+  } while ((err != HB_MC_SUCCESS) || ((max_credits - credits_used) <= 0));
+
+  // Don't need to check for error code since this operation should always be a success since it is not going
+  // over the network
+  err = bp_hb_write_to_mc_bridge(packet);
+  if (err != HB_MC_SUCCESS) {
+    bsg_pr_err("Write to the host request FIFO failed!");
+    return err;
+  }
+
+  return HB_MC_SUCCESS;
+}
+
+/**
+ * Receive a packet from manycore hardware
+ * @param[in] mc       A manycore instance initialized with hb_mc_manycore_init()
+ * @param[in] response A packet into which data should be read
+ * @param[in] timeout  A timeout counter. Unused - set to -1 to wait forever.
+ * @return HB_MC_SUCCESS on success. Otherwise an error code defined in bsg_manycore_errno.h.
+ */
+int hb_mc_platform_receive(hb_mc_manycore_t *mc, hb_mc_packet_t *packet, hb_mc_fifo_rx_t type, long timeout) {
+  hb_mc_platform_t *platform = reinterpret_cast<hb_mc_platform_t *>(mc->platform);
+  const char *typestr = hb_mc_fifo_rx_to_string(type);
+
+  // Timeout is unsupported
+  if (timeout != -1)
+  {
+    manycore_pr_err(mc, "%s: Only a timeout value of -1 is supported\n", __func__);
+    return HB_MC_INVALID;
+  }
+
+  int err;
+  int num_entries = 0;
+  do {
+    err = bp_hb_get_fifo_entries(&num_entries, type);
+  } while ((num_entries == 0) || (err != HB_MC_SUCCESS));
+
+  err = bp_hb_read_from_mc_bridge(packet, type);
+  if (err != HB_MC_SUCCESS) {
+    bsg_pr_err("Read from the %s FIFO did not succeed", typestr);
+    return HB_MC_INVALID;
+  }
+
+  return HB_MC_SUCCESS;
+}
+
+/**
+ * Read the configuration register at an index
+ * @param[in]  mc     A manycore instance initialized with hb_mc_manycore_init()
+ * @param[in]  idx    Configuration register index to access
+ * @param[out] config Configuration value at index
+ * @return HB_MC_SUCCESS on success. Otherwise an error code defined in bsg_manycore_errno.h.
+ */
+int hb_mc_platform_get_config_at(hb_mc_manycore_t *mc, unsigned int idx, hb_mc_config_raw_t *config) {
+  hb_mc_platform_t *platform = reinterpret_cast<hb_mc_platform_t *>(mc->platform);
+  hb_mc_packet_t config_req_pkt, config_resp_pkt;
+  int num_entries = 0;
+  int err;
+
+  if (idx < HB_MC_CONFIG_MAX)
+  {
+    config_req_pkt.request.x_dst = HOST_X_COORD;
+    config_req_pkt.request.y_dst = HOST_Y_COORD;
+    config_req_pkt.request.x_src = BP_HOST_LINK_X;
+    config_req_pkt.request.y_src = BP_HOST_LINK_Y;
+    config_req_pkt.request.op_v2 = HB_MC_PACKET_OP_REMOTE_LOAD;
+    config_req_pkt.request.payload = 0;
+    config_req_pkt.request.reg_id = 0;
+    config_req_pkt.request.addr = HB_MC_HOST_EPA_CONFIG_START + idx;
```

Comment:
Using idx will actually lead to incorrect behavior in hardware because the addr field is actually a word address, not a byte address. It works because the corresponding c code in simulator.cpp treats it as a byte field, but the hardware will not.

In general, when formatting packets it is better to use the packet API. I don't think it's possible here because the configuration struct has not been initialized.

See how the EPA is shifted here: https://github.com/bespoke-silicon-group/bsg_replicant/blob/f29dccb246f17efcfc81948e73ff6532283f998d/libraries/bsg_manycore.cpp#L420

Reply:
Is it possible to put the ROM in a place that can be mapped into BP's address space? That way we wouldn't have to deal with byte and word alignment. The only place I used this mechanism is to retrieve arguments and the configuration.

Reply:
The c code should treat it as a word address since it's going over the network

Reply:
I think the code now reflects the change expected. Correct me if I am wrong, I originally thought the software, like the hardware also assumed that the EPA itself is word addressable.